src/Driver.cpp

```diff
@@ -1,598 +1,530 @@
-        position_string = waitSynchronousString("AT?UP");
-      }
-      return UsblParser::parsePosition(position_string);
+bool Driver::sendRawData(void)
+{
+    if(!queueRawData.empty())
+    {   // TODO Need to check the size of the Raw data?
+        // Maybe some data can be dropped if data's size is too big.
+        std::string buffer = queueRawData.front();
+        std::cout << "Send Raw Data "  << std::endl;
+        bool result = iodrivers_base::Driver::writePacket(reinterpret_cast<const uint8_t*>(buffer.c_str()), buffer.length());
+        if(result)
+            // No need to wait for answer
+            queueRawData.pop();
+        return result;
     }
-//    return current_position;
+    return false;
 }
 
-int Driver::getSystemTime(){
-    sendWithLineEnding("+++AT?UT");
-    return waitSynchronousInt();
+void Driver::sendRawData(std::string const& raw_data)
+{
+    queueRawData.push(raw_data);
 }
-void Driver::open(std::string const& uri, ReverseMode rm){
-    reverse_mode = rm;
-    std::cout << "reverse mode set to " << rm << " (where rm = " << REVERSE_POSITION_SENDER << " enables sending)\n";
-    buffer.resize(50);
-    openURI(uri);
-    if (uri.find("serial") != std::string::npos)
-        interfaceType = SERIAL;
-    else
-        interfaceType = ETHERNET;
-    resetDevice(SEND_BUFFER);
-    requestVersionNumbers();
 
+std::string Driver::fillCommand(std::string const &command)
+{
+	std::stringstream ss;
+	std::string buffer;
+	// Guard Time Escape Sequence (GTES) doesn't require <end-line>
+	// GTES: (1s)<+++>(1s)
+	if(command == "+++")
+	{
+		ss << command << std::flush;
+		if(mode == DATA)
+			sleep(1);
+	}
+	else
+	{
+		if (mode == DATA)
+			// If in DATA, buffer = +++<AT command>
+			ss << "+++" << std::flush;
+		ss << addEndLine(command) << std::flush;
+	}
+	return ss.str();
+}
+
+std::string Driver::addEndLine(std::string const &command)
+{
+	std::stringstream ss;
+	if (interface == SERIAL)
+		ss << command << "\r" << std::flush;
+	else // (interface == ETHERNET)
+		ss << command << "\n" << std::flush;
+	return ss.str();
 }
 
-size_t Driver::read(uint8_t *buffer, size_t size){
-
-	std::cout << "beginning of usbl Driver::read(...)" << std::endl;
-
-	std::cout << "size = " << size << std::endl;
-
-    size_t packet_size = readPacket(buffer, size, 3000, 3000);
-
-    std::string buffer_as_string = std::string(reinterpret_cast<char const*>(buffer));
-    if (packet_size){
-        std::cout << "packet_size: " << packet_size << std::endl;
-        std::cout << "buffer_as_string: " << buffer_as_string << std::endl;
-        if (UsblParser::isPacket(buffer_as_string) > 0){
-            if (handleAsynchronousCommand(buffer_as_string)){
-                packet_size = 0;
+int Driver::extractPacket(uint8_t const* buffer, size_t buffer_size) const
+{
+	std::string buffer_as_string = std::string(reinterpret_cast<char const*>(buffer), buffer_size);
+	buffer_as_string = buffer_as_string.substr(0, buffer_size);
+
+	int ret;
```

Comment:
Don't do that. Define your local variables the closest to where they are needed.